File: python/pov/pov.py
from json import dumps


class Tree:
    def __init__(self, label, children=None):
        self.label = label
        self.children = children if children is not None else []

    def __dict__(self):
        return {self.label: [c.__dict__() for c in sorted(self.children)]}

    def __str__(self, indent=None):
        return dumps(self.__dict__(), indent=indent)

    def __lt__(self, other):
        return self.label < other.label

    def __eq__(self, other):
        return self.__dict__() == other.__dict__()

    def _find_node(self, target_label, parent=None, path=None):
        """Find a node with the given label and return the node and its path from root."""
        if path is None:
            path = []
        
        if self.label == target_label:
            return self, path, parent
        
        for child in self.children:
            result = child._find_node(target_label, self, path + [self])
            if result[0] is not None:
                return result
        
        return None, None, None

    def _create_new_tree(self, target_node, parent_node, visited=None):
        """Create a new tree with the target node as root."""
        if visited is None:
            visited = set()
        
        # Create a copy of the current node
        new_tree = Tree(self.label)
        visited.add(self.label)
        
        # Add all children except the parent
        for child in self.children:
            if child.label != parent_node:
                if child.label not in visited:
                    new_child = child._create_new_tree(target_node, self.label, visited)
                    new_tree.children.append(new_child)
        
        # Add parent as a child if it's not None and not visited
        if parent_node is not None and parent_node != target_node and parent_node not in visited:
            parent = Tree(parent_node)
            new_tree.children.append(parent)
            
        return new_tree

    def from_pov(self, from_node):
        """Reorient the tree with from_node as the new root."""
        # Find the target node
        target, path, parent = self._find_node(from_node)
        
        if target is None:
            raise ValueError("Tree could not be reoriented")
        
        # If the target is already the root, return a copy of the tree
        if parent is None:
            return Tree(self.label, [Tree(child.label, child.children) for child in self.children])
        
        # Create a new tree with the target as root
        new_root = Tree(from_node)
        
        # Add target's children to the new root
        for child in target.children:
            new_root.children.append(Tree(child.label, child.children))
        
        # Build the path from target to original root
        current = parent
        path_to_root = [current]
        
        # Traverse up the path to build the new structure
        for i in range(len(path) - 1, -1, -1):
            node = path[i]
            if node.label == current.label:
                # Find the next node in the path
                if i > 0:
                    current = path[i-1]
                    path_to_root.append(current)
        
        # Build the new tree structure
        current_node = new_root
        for node in path_to_root:
            # Create a new node for the parent
            new_node = Tree(node.label)
            
            # Add all siblings of the target node to the parent
            for child in node.children:
                if child.label != current_node.label and child.label not in [n.label for n in current_node.children]:
                    new_node.children.append(Tree(child.label, child.children))
            
            # Add the parent to the current node's children
            current_node.children.append(new_node)
            current_node = new_node
        
        return new_root

    def path_to(self, from_node, to_node):
        """Find the path from from_node to to_node."""
        # Reorient the tree with from_node as root
        try:
            new_tree = self.from_pov(from_node)
        except ValueError:
            raise ValueError("Tree could not be reoriented")
        
        # Find the path from the new root to to_node
        path = []
        
        def find_path(node, target, current_path):
            if node.label == target:
                return current_path + [node.label]
            
            for child in node.children:
                result = find_path(child, target, current_path + [node.label])
                if result:
                    return result
            
            return None
        
        result = find_path(new_tree, to_node, [])
        
        if result is None:
            raise ValueError("No path found")
        
        return result

File: go/pov/pov.go
package pov

type Tree struct {
	value    string
	children []*Tree
}

// New creates and returns a new Tree with the given root value and children.
func New(value string, children ...*Tree) *Tree {
	return &Tree{
		value:    value,
		children: children,
	}
}

// Value returns the value at the root of a tree.
func (tr *Tree) Value() string {
	return tr.value
}

// Children returns a slice containing the children of a tree.
// There is no need to sort the elements in the result slice,
// they can be in any order.
func (tr *Tree) Children() []*Tree {
	return tr.children
}

// String describes a tree in a compact S-expression format.
// This helps to make test outputs more readable.
// Feel free to adapt this method as you see fit.
func (tr *Tree) String() string {
	if tr == nil {
		return "nil"
	}
	result := tr.Value()
	if len(tr.Children()) == 0 {
		return result
	}
	for _, ch := range tr.Children() {
		result += " " + ch.String()
	}
	return "(" + result + ")"
}

// POV problem-specific functions

// FromPov returns the pov from the node specified in the argument.
func (tr *Tree) FromPov(from string) *Tree {
	// If the tree is nil, return nil
	if tr == nil {
		return nil
	}

	// If the current node is the target, return a copy of the tree
	if tr.Value() == from {
		return tr.clone()
	}

	// Find the path from the root to the target node
	path := tr.findPath(from)
	if path == nil {
		return nil // Target node not found
	}

	// Create a new tree with the target node as the root
	newRoot := &Tree{value: from}

	// Reparent the tree
	current := newRoot
	for i := len(path) - 2; i >= 0; i-- {
		parent := path[i]
		// Create a new node for the parent
		newParent := &Tree{value: parent.Value()}

		// Add all children of the parent except the one in the path
		for _, child := range parent.Children() {
			if i > 0 && child == path[i-1] {
				continue // Skip the child that's in the path
			}
			if child.Value() == current.Value() {
				continue // Skip the current node
			}
			newParent.children = append(newParent.children, child.clone())
		}

		// Add the parent as a child of the current node
		current.children = append(current.children, newParent)
		current = newParent
	}

	// If the target node has children, add them to the new root
	targetNode := path[len(path)-1]
	for _, child := range targetNode.Children() {
		newRoot.children = append(newRoot.children, child.clone())
	}

	return newRoot
}

// PathTo returns the shortest path between two nodes in the tree.
func (tr *Tree) PathTo(from, to string) []string {
	// First, reorient the tree from the perspective of 'from'
	fromPov := tr.FromPov(from)
	if fromPov == nil {
		return nil // 'from' node not found
	}

	// Find the path from the new root ('from') to 'to'
	path := []string{from} // Start with 'from'
	if from == to {
		return path // If 'from' and 'to' are the same, return just 'from'
	}

	// Find the path from 'from' to 'to'
	found := fromPov.findPathValues(to, &path)
	if !found {
		return nil // 'to' node not found
	}

	return path
}

// Helper function to find a path from the root to a node with the given value
func (tr *Tree) findPath(value string) []*Tree {
	if tr.Value() == value {
		return []*Tree{tr}
	}

	for _, child := range tr.Children() {
		if path := child.findPath(value); path != nil {
			return append([]*Tree{tr}, path...)
		}
	}

	return nil
}

// Helper function to find a path from the root to a node with the given value
// and store the values in the provided path slice
func (tr *Tree) findPathValues(value string, path *[]string) bool {
	if tr.Value() == value {
		return true
	}

	for _, child := range tr.Children() {
		*path = append(*path, child.Value())
		if child.findPathValues(value, path) {
			return true
		}
		*path = (*path)[:len(*path)-1] // Backtrack
	}

	return false
}

// Helper function to create a deep copy of a tree
func (tr *Tree) clone() *Tree {
	if tr == nil {
		return nil
	}

	newTree := &Tree{value: tr.Value()}
	for _, child := range tr.Children() {
		newTree.children = append(newTree.children, child.clone())
	}

	return newTree
}
